### PR TITLE
Add telemetry publisher about Knob values per Job request

### PR DESCRIPTION
**Pull request by ismayilov-ismayil**
Issue: Knobs value might be overridden by job request and it's hard to track knob value per job
Sln: Added telemetry publisher for knob values

[AB#2205888](https://dev.azure.com/mseng/b924d696-3eae-4116-8443-9a18392d8544/_workitems/edit/2205888)

### CODE REVIEW

File: src/Agent.Worker/JobExtension.cs [hunk 1]
```diff
@@ -726,6 +727,33 @@ private void OutputSetupInfo(IExecutionContext context)
             }
         }
 
+        private void PublishKnobsInfo(IExecutionContext jobContext)
+        {
+            Debugger.Launch();
```
[the remark this answers is not in the document]
Comment (ismayilov-ismayil) [reply src/Agent.Worker/JobExtension.cs]: done

File: src/Agent.Worker/JobExtension.cs [hunk 2]
```diff
@@ -726,6 +727,33 @@ private void OutputSetupInfo(IExecutionContext context)
...
+                    var stringValue = value.AsString();
+                    if (knob is SecretKnob)
+                    {
+                        HostContext.SecretMasker.AddValue(stringValue, $"JobExtension_InitializeJob_{knob.Name}");
+                    }
+                    telemetryData.Add($"{knob.Name}-{value.Source.GetDisplayString()}", stringValue);
+                }
+            }
+
+            PublishTelemetry(jobContext, telemetryData, "KnobsStatus");
```
[the remark this answers is not in the document]
Comment (ismayilov-ismayil) [reply src/Agent.Worker/JobExtension.cs:753]: PublishTelemetry method contains error handler

File: src/Agent.Worker/JobExtension.cs [hunk 3]
```diff
@@ -726,6 +727,32 @@ private void OutputSetupInfo(IExecutionContext context)
...
+            foreach (var knob in Knob.GetAllKnobsFor<AgentKnobs>())
+            {
+                var value = knob.GetValue(jobContext);
+                if (value.Source.GetType() != typeof(BuiltInDefaultKnobSource))
+                {
+                    var stringValue = value.AsString();
+                    if (knob is SecretKnob)
+                    {
+                        HostContext.SecretMasker.AddValue(stringValue, $"JobExtension_InitializeJob_{knob.Name}");
+                    }
```
Comment (KonstantinTyukalov) [inline src/Agent.Worker/JobExtension.cs:748]: Isn't this made already?
https://github.com/microsoft/azure-pipelines-agent/blob/f60062e975aab1c9b0a4bc9ec4a87b3e61c081e5/src/Agent.Worker/JobExtension.cs#L186-L189

File: src/Agent.Worker/JobExtension.cs [hunk 4]
```diff
@@ -726,6 +727,32 @@ private void OutputSetupInfo(IExecutionContext context)
...
         }
 
+        private void PublishKnobsInfo(IExecutionContext jobContext)
+        {
+            string jobId = jobContext?.Variables?.System_JobId?.ToString() ?? string.Empty;
+
+            var telemetryData = new Dictionary<string, string>()
+            {
+                { "JobId", jobId }
+            };
```
Comment (KonstantinTyukalov) [inline src/Agent.Worker/JobExtension.cs:737]: Nit: it's probably better to post null instead of empty string as it's easier to filter later

**Review by KonstantinTyukalov — APPROVED** [review]

File: src/Agent.Worker/JobExtension.cs (hunk 4 above)
Comment (ismayilov-ismayil) [reply src/Agent.Worker/JobExtension.cs]: done

**Review by kirill-ivlev — APPROVED** [review]

**Review by DergachevE — APPROVED** [review]

**Pull request merged**